src/main/com/fulcrologic/fulcro/algorithms/normalize.cljc

```diff
@@ -18,74 +20,193 @@
+              (let [[join-key subquery] (util/join-entry expr)
+                    recursive?  (util/recursion? subquery)
+                    union-entry (if (util/union? expr) subquery union-seen)
+                    subquery    (if recursive?
+                                  (if-not (nil? union-seen)
+                                    union-seen
+                                    query)
+                                  subquery)
+                    class       (-> subquery meta :component)
+                    v           (get data join-key)]
+                (cond
+                  ;; graph loop: db->tree leaves ident in place
+                  (and recursive? (eql/ident? v)) (recur (next q) ret)
+                  ;; normalize one
+                  (map? v)
+                  (let [x (normalize-with-friendly-errors subquery v tables union-entry transform)]
+                    (if-not (or (nil? class) (not (has-ident? class)))
+                      (let [i (get-ident class x)]
+                        ;; Why don't we simply `update-in i ..` as we do below in normalize many?! Incidental?
+                        (swap! tables upsert-ident i x)
+                        (recur (next q) (assoc ret join-key i)))
+                      (recur (next q) (assoc ret join-key x))))
+
+                  ;; normalize many
+                  (and (vector? v) (not (eql/ident? v)) (not (eql/ident? (first v))))
+                  (let [xs (into [] (map #(normalize-with-friendly-errors subquery % tables union-entry transform)) v)]
+                    (if-not (or (nil? class) (not (has-ident? class)))
+                      (let [is (into [] (map #(get-ident class %)) xs)]
+                        ;; union + normal case
+                        (swap! tables
+                               (fn [tables']
+                                 (reduce (fn merge-to-client-db [m [i x]] (upsert-ident m i x))
+                                         tables' (map vector is xs))))
+                        (recur (next q) (assoc ret join-key is)))
+                      (recur (next q) (assoc ret join-key xs))))
+
+                  ;; missing key
+                  (nil? v)
+                  (recur (next q) ret)
+
+                  ;; can't handle
+                  :else (recur (next q) (assoc ret join-key v))))
+              (let [k (if (seq? expr) (first expr) expr)
+                    v (get data k)]
+                (if (nil? v)
+                  (recur (next q) ret)
+                  (recur (next q) (assoc ret k v))))))
+          ret)))))
+
+(defn- better-normalize* [query data tables union-seen transform]
+  (try
+    (normalize* query data tables union-seen transform)
+    (catch #?(:clj Exception :cljs :default) e
+      (normalize-with-friendly-errors query data tables union-seen transform)
+      (log/warn "Unexpectedly, normalize* failed but normalize-with-friendly-errors did not. Please report it (with the inputs) to Fulcro")
```

Comment:
I would think the ideal thing to do would be report error, then issue a final error that says "normalize failed, see above errors" and maybe includes the query and data that was provided...but don't throw again. A stack trace is almost certainly non-helpful since it will come from the low-level transaction system, not your code.

Reply:
Ok! Pushed a new version.

I still throw an exception in `upsert-ident` just to fail fast (and get the info faster to the developer) but changed this place to only log.